doc/serdi.1
.\" Copyright 2011-2022 David Robillard <d@drobilla.net>
.\" SPDX-License-Identifier: ISC
.Dd April 30, 2023
.Dt SERDI 1
.Os Serd 0.31.5
.Sh NAME
.Nm serdi
.Nd read and write RDF syntax
.Sh SYNOPSIS
.Nm serdi
.Op Fl abefhlqv
.Op Fl c Ar prefix
.Op Fl i Ar syntax
.Op Fl o Ar syntax
.Op Fl p Ar prefix
.Op Fl r Ar root
.Op Fl s Ar string
.Ar input
.Op Ar base_uri
.Sh DESCRIPTION
.Nm
is a fast command-line utility for streaming and processing RDF data.
It reads an RDF document and writes the data to stdout,
possibly transformed and/or in a different syntax.
By default,
the input syntax is guessed from the file extension,
and output is written in NTriples or NQuads.
.Pp
.Nm
can be used to check for syntax errors,
convert from one syntax to another,
pretty-print documents,
or transform URIs and blank node IDs.
.Pp
The options are as follows:
.Bl -tag -width 3n
.It Fl a
Write ASCII output.
If this is enabled, all non-ASCII characters will be escaped, even if the output syntax allows them to be written in UTF-8.
.It Fl b
Bulk output writing.
If this is enabled, output will be written a page at a time, rather than a byte at a time.
.It Fl c Ar prefix
Chop
.Ar prefix
from matching blank node IDs.
This is typically used to revert the effects of
.Fl p .
For example, with
.Ar prefix
.Dq doc01 ,
the blank node
.Li _:doc01b42
will be emitted as
.Li _:b42 .
.It Fl e
Eat input one character at a time, rather than a page at a time which is the default.
This is useful when reading from a pipe since output will be generated immediately as input arrives, rather than waiting until an entire page of input has arrived.
With this option serdi uses one page less memory, but will likely be significantly slower.
.It Fl f
Fast and loose URI mode:
preserve full URIs (without qualifying or making relative),
and pass prefixed names through as-is.
.It Fl h
Print the command line options.
.It Fl i Ar syntax
Read input as
.Ar syntax .
Case is ignored, valid values are:
.Dq NQuads ,
.Dq NTriples ,
.Dq TriG ,
.Dq Turtle .
.It Fl l
Lax (non-strict) parsing.
If this is enabled, recoverable syntax errors will print a warning, but parsing will proceed starting at the next statement if possible.
Note that data may be lost when using this option.
.It Fl o Ar syntax
Write output as
.Ar syntax .
Case is ignored, valid values are:
.Dq NQuads ,
.Dq NTriples ,
.Dq TriG ,
.Dq Turtle .
.It Fl p Ar prefix
Add
.Ar prefix
to blank node IDs.
This can be used to avoid clashes between blank node IDs in input documents.
The effects can be reversed in a later run with
.Fl c .
For example, with
.Ar prefix
.Dq doc01 ,
the blank node
.Li _:b42
will be emitted as
.Li _:doc01b42 .
.It Fl q
Suppress all output except data.
.It Fl r Ar root
Keep relative URIs within a
.Ar root
URI.
This will avoid creating any relative URI references with leading path segments like
.Pa ../
that enter a parent of
.Ar root .
.Pp
For example,
if
.Pa /home/you/file.ttl
is written to the file
.Pa /home/me/output.ttl
using the destination's base URI,
then it could be written as
.Li <../you/file.ttl> .
Setting
.Fl r Li file:///home/me/
would prevent references from
.Dq escaping
like this,
so the above would instead be written as
.Li <file:///home/you/file.ttl> ,
since it can't be expressed relative to the root URI.
.Pp
This is useful for keeping relative references within some directory.
.It Fl s Ar string
Parse
.Ar string
input instead of a file (terminates options).
.It Fl v
Display version information and exit.
.El
.Sh EXIT STATUS
.Nm
exits with a status of 0, or non-zero if an error occurred.
.Sh EXAMPLES
.Bl -tag -width 3n
.It Pretty-print a document:
.Nm Fl o
.Ar turtle
.Pa file.ttl
>
.Pa out.ttl
.It Print any errors:
.Nm Pa file.ttl
>
.Pa /dev/null
.El
.Sh SEE ALSO
.Bl -item -compact
.It
.Lk http://drobilla.net/software/serd/
.It
.Lk http://gitlab.com/drobilla/serd/
.El
.Sh STANDARDS
.Bl -item
.It
.Rs
.%A W3C
.%T RDF 1.1 NQuads
.%D February 2014
.Re
.Lk https://www.w3.org/TR/n-quads/
.It
.Rs
.%A W3C
.%D February 2014
.%T RDF 1.1 NTriples
.Re
.Lk https://www.w3.org/TR/n-triples/
.It
.Rs
.%A W3C
.%T RDF 1.1 TriG
.%D February 2014
.Re
.Lk https://www.w3.org/TR/trig/
.It
.Rs
.%A W3C
.%D February 2014
.%T RDF 1.1 Turtle
.Re
.Lk https://www.w3.org/TR/turtle/
.El
.Sh AUTHORS
.Nm
is a part of serd, by
.An David Robillard
.Mt d@drobilla.net .


NEWS
serd (0.31.5) unstable; urgency=medium

  * Add Windows path separator support to serd_node_new_file_uri()
  * Add long "help" and "version" options to serdi
  * Add options to disable html or singlehtml documentation
  * Add serd_reader_skip_until_byte() to public API
  * Allow SERD_API to be defined by the user
  * Avoid creating test files in the current directory
  * Avoid using ASCII grave as a quote
  * Check for POSIX features with the build system
  * Clean up and improve test suite
  * Clean up code
  * Fix crash when trying to read chunks without starting
  * Fix hang when skipping an error at EOF when lax parsing
  * Fix incorrect parsing of strange quote escape patterns
  * Fix possible hang when writing nested Turtle lists
  * Fix potential memory leaks when a write is aborted
  * Fix relative URI creation
  * Gracefully handle bad characters in Turtle blank node syntax
  * Gracefully handle bad characters in Turtle datatype syntax
  * Improve TriG pretty-printing and remove trailing newlines
  * Improve pretty-printing of lists and inline subjects
  * Improve serdi man page
  * Improve writer error handling
  * Make URI writing stricter by default
  * Make serd_reader_read_chunk() work with NQuads
  * Override pkg-config dependency within meson
  * Remove junk files from documentation install
  * Remove support for writing Turtle named inline nodes extension
  * Replace duplicated dox_to_sphinx script with sphinxygen dependency
  * Test header for warnings more strictly
  * Update standard test suites

 -- David Robillard <d@drobilla.net>  Fri, 12 May 2023 02:42:30 +0000

serd (0.30.16) stable; urgency=medium

  * Add SERD_STATIC to pkg-config Cflags for static-only builds
  * Adopt REUSE machine-readable licensing standard
  * Allow programs to be used from subproject
  * Fix spelling mistake in serdi man page

 -- David Robillard <d@drobilla.net>  Fri, 09 Sep 2022 03:06:45 +0000

serd (0.30.14) stable; urgency=medium

  * Fix memory consumption when reading documents
  * Switch to Meson build system
  * Update README and project metadata

 -- David Robillard <d@drobilla.net>  Sun, 10 Jul 2022 16:58:43 +0000

serd (0.30.12) stable; urgency=medium

  * Fix warnings and build issues with clang 13 and VS 2019
  * Fix writing long literals with triple quotes
  * Improve documentation style
  * Support combining several BSD-style command line flags in serdi
  * Write statements with invalid URI characters in lax mode

 -- David Robillard <d@drobilla.net>  Thu, 26 May 2022 22:47:35 +0000

serd (0.30.10) stable; urgency=medium

  * Add fallback configuration if documentation theme is unavailable
  * Fix SERD_DISABLE_DEPRECATED
  * Fix building for older MacOS versions on newer MacOS
  * Fix documentation installation directory

 -- David Robillard <d@drobilla.net>  Wed, 20 Jan 2021 15:55:35 +0000

serd (0.30.8) stable; urgency=medium

  * Deprecate serd_uri_to_path()
  * Don't install API man pages
  * Fix potential memory error when serialising URIs
  * Move headers to an include directory
  * Refuse to write relative URI references to NTriples
  * Remove the need for a generated configuration header
  * Remove use of C character class functions that may use locale
  * Split up and reorganize unit tests
  * Use aligned allocation via C11 or Windows API where possible

 -- David Robillard <d@drobilla.net>  Thu, 07 Jan 2021 21:19:36 +0000

serd (0.30.6) stable; urgency=medium

  * Fix various minor warnings and other code quality issues
  * Improve reader error handling
  * Make all free functions tolerant of NULL
  * Parse prefixed names more strictly

 -- David Robillard <d@drobilla.net>  Sun, 27 Sep 2020 10:51:38 +0000

serd (0.30.4) stable; urgency=medium

  * Fix EOF handling while reading in bulk or from strings
  * Fix lax handling of string errors
  * Fix reading from a null-delimited socket

 -- David Robillard <d@drobilla.net>  Sun, 26 Apr 2020 16:04:05 +0000

serd (0.30.2) stable; urgency=medium

  * Fix GCC 4 build
  * Fix colliding blank nodes when parsing TriG
  * Fix missing parse error messages
  * Fix parsing "a" abbreviation without padding whitespace
  * Fix parsing TriG graphs with several squashed trailing dots
  * Fix resolving some URIs against base URIs with no trailing slash
  * Improve build system and CI integration
  * Improve documentation

 -- David Robillard <d@drobilla.net>  Sun, 20 Oct 2019 22:28:01 +0000

serd (0.30.0) stable; urgency=medium

  * Add NTriples test suite
  * Add basic error tolerant parsing for line-based formats
  * Add serd_free for freeing memory allocated by serd
  * Add serdi option to write ASCII output
  * Clarify errors returned by serd_env_expand()
  * Clean up testing code
  * Fix building with MSVC
  * Fix length of base64 encoded nodes
  * Fix parsing local names that end with escaped dots
  * Fix reported error when reading statements with only a blank node
  * Improve relative URI serialisation
  * Make serdi guess input syntax from extension if unspecified
  * Make serdi syntax options case-insensitive
  * Report errors for all failing test cases
  * Support strict parsing of prefixed names
  * Upgrade to waf 2.0.9 and autowaf a328983
  * Use more efficient stack allocation and growth policies

 -- David Robillard <d@drobilla.net>  Sun, 22 Jul 2018 18:34:09 +0000

serd (0.28.0) stable; urgency=medium

  * Add serd_node_from_substring()
  * Add support for reading from a user provided callback
  * Fix hangs when reading corrupt UTF-8
  * Fix parsing of hex escapes in file URIs (thanks Johannes Mueller)
  * Fix strict parsing of abolute URI schemes
  * Gracefully handle applications that write corrupt UTF-8

 -- David Robillard <d@drobilla.net>  Fri, 21 Jul 2017 06:28:47 +0000

serd (0.26.0) stable; urgency=medium

  * Add support for TriG and NQuads
  * Add support for braindead SPARQL PREFIX and BASE directives

 -- David Robillard <d@drobilla.net>  Tue, 10 Jan 2017 01:10:35 +0000


src/serdi.c
// Copyright 2011-2023 David Robillard <d@drobilla.net>
// SPDX-License-Identifier: ISC

#include "serd_config.h"
#include "string_utils.h"

#include "serd/attributes.h"
#include "serd/env.h"
#include "serd/error.h"
#include "serd/node.h"
#include "serd/reader.h"
#include "serd/sink.h"
#include "serd/status.h"
#include "serd/syntax.h"
#include "serd/uri.h"
#include "serd/version.h"
#include "serd/writer.h"

#ifdef _WIN32
#  ifdef _MSC_VER
#    define WIN32_LEAN_AND_MEAN
#  endif
#  include <fcntl.h>
#  include <io.h>
#endif

#if USE_POSIX_FADVISE && USE_FILENO
#  include <fcntl.h>
#endif

#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SERDI_ERROR(msg) fprintf(stderr, "serdi: " msg)
#define SERDI_ERRORF(fmt, ...) fprintf(stderr, "serdi: " fmt, __VA_ARGS__)

typedef struct {
  SerdSyntax  syntax;
  const char* name;
  const char* extension;
} Syntax;

static const Syntax syntaxes[] = {{SERD_TURTLE, "turtle", ".ttl"},
                                  {SERD_NTRIPLES, "ntriples", ".nt"},
                                  {SERD_NQUADS, "nquads", ".nq"},
                                  {SERD_TRIG, "trig", ".trig"},
                                  {(SerdSyntax)0, NULL, NULL}};

static SerdSyntax
get_syntax(const char* const name)
{
  for (const Syntax* s = syntaxes; s->name; ++s) {
    if (!serd_strncasecmp(s->name, name, strlen(name))) {
      return s->syntax;
    }
  }

  SERDI_ERRORF("unknown syntax '%s'\n", name);
  return (SerdSyntax)0;
}

static SERD_PURE_FUNC SerdSyntax
guess_syntax(const char* const filename)
{
  const char* ext = strrchr(filename, '.');
  if (ext) {
    for (const Syntax* s = syntaxes; s->name; ++s) {
      if (!serd_strncasecmp(s->extension, ext, strlen(ext))) {
        return s->syntax;
      }
    }
  }

  return (SerdSyntax)0;
}

static int
print_version(void)
{
  printf("serdi %d.%d.%d <http://drobilla.net/software/serd>\n",
         SERD_MAJOR_VERSION,
         SERD_MINOR_VERSION,
         SERD_MICRO_VERSION);

  printf("Copyright 2011-2023 David Robillard <d@drobilla.net>.\n"
         "License ISC: <https://spdx.org/licenses/ISC>.\n"
         "This is free software; you are free to change and redistribute it."
         "\nThere is NO WARRANTY, to the extent permitted by law.\n");

  return 0;
}

static int
print_usage(const char* const name, const bool error)
{
  static const char* const description =
    "Read and write RDF syntax.\n"
    "Use - for INPUT to read from standard input.\n\n"
    "  -a           Write ASCII output.\n"
    "  -b           Write output in blocks for performance.\n"
    "  -c PREFIX    Chop PREFIX from matching blank node IDs.\n"
    "  -e           Eat input one character at a time.\n"
    "  -f           Fast and loose URI pass-through.\n"
    "  -h           Display this help and exit.\n"
    "  -i SYNTAX    Input syntax: turtle/ntriples/trig/nquads.\n"
    "  -l           Lax (non-strict) parsing.\n"
    "  -o SYNTAX    Output syntax: turtle/ntriples/nquads.\n"
    "  -p PREFIX    Add PREFIX to blank node IDs.\n"
    "  -q           Suppress all output except data.\n"
    "  -r ROOT_URI  Keep relative URIs within ROOT_URI.\n"
    "  -s INPUT     Parse INPUT as string (terminates options).\n"
    "  -v           Display version information and exit.\n";

  FILE* const os = error ? stderr : stdout;
  fprintf(os, "%s", error ? "\n" : "");
  fprintf(os, "Usage: %s [OPTION]... INPUT [BASE_URI]\n", name);
  fprintf(os, "%s", description);
  return error ? 1 : 0;
}

static int
missing_arg(const char* const name, const char opt)
{
  SERDI_ERRORF("option requires an argument -- '%c'\n", opt);
  return print_usage(name, true);
}

static SerdStatus
quiet_error_sink(void* const handle, const SerdError* const e)
{
  (void)handle;
  (void)e;
  return SERD_SUCCESS;
}

static FILE*
serd_fopen(const char* const path, const char* const mode)
{
  FILE* fd = fopen(path, mode);
  if (!fd) {
    SERDI_ERRORF("failed to open file %s (%s)\n", path, strerror(errno));
    return NULL;
  }

#if USE_POSIX_FADVISE && USE_FILENO
  (void)posix_fadvise(
    fileno(fd), 0, 0, POSIX_FADV_SEQUENTIAL | POSIX_FADV_NOREUSE);
#endif

  return fd;
}

static SerdWriterFlags
choose_style(const SerdSyntax input_syntax,
             const SerdSyntax output_syntax,
             const bool       ascii,
             const bool       bulk_write,
             const bool       full_uris,
             const bool       lax)
{
  SerdWriterFlags writer_flags = 0U;
  if (output_syntax == SERD_NTRIPLES || ascii) {
    writer_flags |= SERD_WRITE_ASCII;
  } else if (output_syntax == SERD_TURTLE) {
    writer_flags |= SERD_WRITE_ABBREVIATED;
    if (!full_uris) {
      writer_flags |= SERD_WRITE_CURIED;
    }
  }

  if ((input_syntax == SERD_TURTLE || input_syntax == SERD_TRIG) ||
      (writer_flags & SERD_WRITE_CURIED)) {
    // Base URI may change and/or we're abbreviating URIs, so must resolve
    writer_flags |= SERD_WRITE_RESOLVED;
  }

  if (bulk_write) {
    writer_flags |= SERD_WRITE_BULK;
  }

  if (!lax) {
    writer_flags |= SERD_WRITE_STRICT;
  }

  return writer_flags;
}

int
main(int argc, char** argv)
{
  const char* const prog = argv[0];

  FILE*       in_fd         = NULL;
  SerdSyntax  input_syntax  = (SerdSyntax)0;
  SerdSyntax  output_syntax = (SerdSyntax)0;
  bool        from_file     = true;
  bool        ascii         = false;
  bool        bulk_read     = true;
  bool        bulk_write    = false;
  bool        full_uris     = false;
  bool        lax           = false;
  bool        quiet         = false;
  const char* in_name       = NULL;
  const char* add_prefix    = NULL;
  const char* chop_prefix   = NULL;
  const char* root_uri      = NULL;
  int         a             = 1;
  for (; a < argc && from_file && argv[a][0] == '-'; ++a) {
    if (argv[a][1] == '\0') {
      in_name = (const char*)"(stdin)";
      in_fd   = stdin;
      break;
    }

    if (!strcmp(argv[a], "--help")) {
      return print_usage(prog, false);
    }

    if (!strcmp(argv[a], "--version")) {
      return print_version();
    }

    for (int o = 1; argv[a][o]; ++o) {
      const char opt = argv[a][o];

      if (opt == 'a') {
        ascii = true;
      } else if (opt == 'b') {
        bulk_write = true;
      } else if (opt == 'e') {
        bulk_read = false;
      } else if (opt == 'f') {
        full_uris = true;
      } else if (opt == 'h') {
        return print_usage(prog, false);
      } else if (opt == 'l') {
        lax = true;
      } else if (opt == 'q') {
        quiet = true;
      } else if (opt == 'v') {
        return print_version();
      } else if (opt == 's') {
        in_name   = "(string)";
        from_file = false;
        break;
      } else if (opt == 'c') {
        if (argv[a][o + 1] || ++a == argc) {
          return missing_arg(prog, 'c');
        }

        chop_prefix = argv[a];
        break;
      } else if (opt == 'i') {
        if (argv[a][o + 1] || ++a == argc) {
          return missing_arg(prog, 'i');
        }

        if (!(input_syntax = get_syntax(argv[a]))) {
          return print_usage(prog, true);
        }
        break;
      } else if (opt == 'o') {
        if (argv[a][o + 1] || ++a == argc) {
          return missing_arg(prog, 'o');
        }

        if (!(output_syntax = get_syntax(argv[a]))) {
          return print_usage(prog, true);
        }
        break;
      } else if (opt == 'p') {
        if (argv[a][o + 1] || ++a == argc) {
          return missing_arg(prog, 'p');
        }

        add_prefix = argv[a];
        break;
      } else if (opt == 'r') {
        if (argv[a][o + 1] || ++a == argc) {
          return missing_arg(prog, 'r');
        }

        root_uri = argv[a];
        break;
      } else {
        SERDI_ERRORF("invalid option -- '%s'\n", argv[a] + 1);
        return print_usage(prog, true);
      }
    }
  }

  if (a == argc) {
    SERDI_ERROR("missing input\n");
    return print_usage(prog, true);
  }

#ifdef _WIN32
  _setmode(_fileno(stdin), _O_BINARY);
  _setmode(_fileno(stdout), _O_BINARY);
#endif

  char*       input_path = NULL;
  const char* input      = (const char*)argv[a++];
  if (from_file) {
    in_name = in_name ? in_name : input;
    if (!in_fd) {
      if (!strncmp(input, "file:", 5)) {
        input_path = serd_file_uri_parse(input, NULL);
        input      = input_path;
      }
      if (!input || !(in_fd = serd_fopen(input, "rb"))) {
        return 1;
      }
    }
  }

  if (!input_syntax && !(input_syntax = guess_syntax(in_name))) {
    input_syntax = SERD_TRIG;
  }

  if (!output_syntax) {
    output_syntax =
      ((input_syntax == SERD_TURTLE || input_syntax == SERD_NTRIPLES)
         ? SERD_NTRIPLES
         : SERD_NQUADS);
  }

  const SerdWriterFlags writer_flags = choose_style(
    input_syntax, output_syntax, ascii, bulk_write, full_uris, lax);

  SerdURIView base_uri = SERD_URI_NULL;
  SerdNode    base     = SERD_NODE_NULL;
  if (a < argc) { // Base URI given on command line
    base = serd_node_new_uri_from_string((const char*)argv[a], NULL, &base_uri);
  } else if (from_file && in_fd != stdin) { // Use input file URI
    base = serd_node_new_file_uri(input, NULL, &base_uri);
  }

  FILE* const    out_fd = stdout;
  SerdEnv* const env    = serd_env_new(&base);

  SerdWriter* const writer = serd_writer_new(
    output_syntax, writer_flags, env, &base_uri, serd_file_sink, out_fd);

  SerdReader* const reader =
    serd_reader_new(input_syntax,
                    writer,
                    NULL,
                    (SerdBaseFunc)serd_writer_set_base_uri,
                    (SerdPrefixFunc)serd_writer_set_prefix,
                    (SerdStatementFunc)serd_writer_write_statement,
                    (SerdEndFunc)serd_writer_end_anon);

  serd_reader_set_strict(reader, !lax);
  if (quiet) {
    serd_reader_set_error_sink(reader, quiet_error_sink, NULL);
    serd_writer_set_error_sink(writer, quiet_error_sink, NULL);
  }

  SerdNode root = serd_node_from_string(SERD_URI, root_uri);
  serd_writer_set_root_uri(writer, &root);
  serd_writer_chop_blank_prefix(writer, chop_prefix);
  serd_reader_add_blank_prefix(reader, add_prefix);

  SerdStatus st = SERD_SUCCESS;
  if (!from_file) {
    st = serd_reader_read_string(reader, input);
  } else if (bulk_read) {
    st = serd_reader_read_file_handle(reader, in_fd, in_name);
  } else {
    st = serd_reader_start_stream(reader, in_fd, in_name, false);
    while (!st) {
      st = serd_reader_read_chunk(reader);
    }
    serd_reader_end_stream(reader);
  }

  serd_reader_free(reader);
  serd_writer_finish(writer);
  serd_writer_free(writer);
  serd_env_free(env);
  serd_node_free(&base);
  free(input_path);

  if (from_file) {
    fclose(in_fd);
  }

  if (fclose(out_fd)) {
    perror("serdi: write error");
    st = SERD_BAD_STREAM;
  }

  return (st > SERD_FAILURE) ? 1 : 0;
}
